rust/worker/src/blockstore/provider.rs

```diff
@@ -0,0 +1,98 @@
+use super::types::Blockfile;
+use super::types::{HashMapBlockfile, KeyType, ValueType};
+use crate::errors::ChromaError;
+use parking_lot::RwLock;
+use std::collections::HashMap;
+use std::sync::Arc;
+use thiserror::Error;
+
+// =================== Interfaces ===================
+
+/// A trait for opening and creating blockfiles
+/// # Methods
+/// - new: Create a new instance of the blockfile provider. A blockfile provider returns a Box<dyn Blockfile> of a given type.
+/// Currently, we support HashMap and Arrow-backed blockfiles.
+/// - open: Open a blockfile at the given path, returning a Box<dyn Blockfile> and error if it does not exist
+/// - create: Create a new blockfile at the given path, returning a Box<dyn Blockfile> and error if it already exists
+/// # Example
+/// ```ignore (TODO: This example is not runnable from outside the crate it seems. Fix this. Ignore for now.)
+/// use crate::blockstore::provider::HashMapBlockfileProvider;
+/// use crate::blockstore::types::{KeyType, ValueType};
+/// let mut provider = HashMapBlockfileProvider::new();
+/// let blockfile = provider.create("test", KeyType::String, ValueType::Int32Array);
+/// ```
+pub(crate) trait BlockfileProvider {
+    fn new() -> Self;
+    fn open(&self, path: &str) -> Result<Box<dyn Blockfile>, Box<OpenError>>;
+    fn create(
+        &mut self,
+        path: &str,
+        key_type: KeyType,
+        value_type: ValueType,
+    ) -> Result<Box<dyn Blockfile>, Box<CreateError>>;
+}
+
+/// A BlockFileProvider that creates HashMapBlockfiles (in-memory blockfiles used for testing).
+/// It bookkeeps the blockfiles locally.
```

Comment:
`bookkeep` (and conjugations/portmanteaus) is the only word in the english language with three double-letters in a row

Reply:
woah